src/share/cxx/CamForcing.cpp

```diff
@@ -0,0 +1,190 @@
+      });
+  Kokkos::parallel_for(
+      "state velocity forcing",
+      Kokkos::RangePolicy<ExecSpace>(0, num_e * 2 * NP * NP * NUM_LEV),
+      KOKKOS_LAMBDA(const int & idx) {
+        const int ie = (((idx / NUM_LEV) / NP) / NP) / 2;
+        const int dim = (((idx / NUM_LEV) / NP) / NP) % 2;
+        const int igp = ((idx / NUM_LEV) / NP) % NP;
+        const int jgp = (idx / NUM_LEV) % NP;
+        const int k = idx % NUM_LEV;
+        v(ie, np1, dim, igp, jgp, k) += dt * f_m(ie, dim, igp, jgp, k);
+      });
+}
+
+void tracer_forcing(
+    const ExecViewUnmanaged<const Scalar * [QSIZE_D][NP][NP][NUM_LEV]> &f_q,
+    const HybridVCoord &hvcoord, const TimeLevel &tl, const int &num_q,
+    const MoistDry &moisture, const double &dt,
+    const ExecViewManaged<Real * [NUM_TIME_LEVELS][NP][NP]> &ps_v,
+    const ExecViewManaged<
+        Scalar * [Q_NUM_TIME_LEVELS][QSIZE_D][NP][NP][NUM_LEV]> &qdp,
+    const ExecViewManaged<Scalar * [QSIZE_D][NP][NP][NUM_LEV]> &Q) {
+
+  const int num_e = ps_v.extent_int(0);
+
+  const auto policy = Homme::get_default_team_policy<ExecSpace>(num_e);
+
+  Kokkos::parallel_for(
+      "tracer qdp forcing",
+      Kokkos::RangePolicy<ExecSpace>(0, num_e * num_q * NP * NP * NUM_LEV),
+      KOKKOS_LAMBDA(const int & idx) {
+        const int ie = (((idx / NUM_LEV) / NP) / NP) / num_q;
+        const int iq = (((idx / NUM_LEV) / NP) / NP) % num_q;
+        const int igp = ((idx / NUM_LEV) / NP) % NP;
+        const int jgp = (idx / NUM_LEV) % NP;
+        const int k = idx % NUM_LEV;
+        Scalar v1 = f_q(ie, iq, igp, jgp, k);
+        Scalar &qdp_s = qdp(ie, tl.np1_qdp, iq, igp, jgp, k);
+        VECTOR_SIMD_LOOP
+        for (int vlev = 0; vlev < VECTOR_SIZE; ++vlev) {
+          if (qdp_s[vlev] + v1[vlev] * dt < 0.0 && v1[vlev] < 0.0) {
+            if (qdp_s[vlev] < 0.0) {
+              v1[vlev] = 0.0;
+            } else {
+              v1[vlev] = -qdp_s[vlev] / dt;
+            }
+          }
+          qdp_s[vlev] += v1[vlev];
+        }
+      });
+
+  if (moisture == MoistDry::MOIST) {
+    // Remove the m_fq_ps_v buffer since it's not actually needed.
+    // Instead apply the forcing to m_ps_v directly
+    // Bonus - one less parallel reduce in dry cases!
+
+    // This conserves the dry mass in the process of forcing tracer 0
+    Kokkos::parallel_for("tracer FQ_PS forcing", policy,
+                         KOKKOS_LAMBDA(const TeamMember & team) {
+      KernelVariables kv(team);
+      const int &ie = kv.team_idx;
+
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(kv.team, NP * NP),
+                           [&](const int pt_idx) {
+        const int igp = pt_idx / NP;
+        const int jgp = pt_idx % NP;
+
+        Real ps_v_forcing = 0.0;
+
+        Dispatch<ExecSpace>::parallel_reduce(
+            kv.team, Kokkos::ThreadVectorRange(kv.team, NUM_PHYSICAL_LEV),
+            [&](const int &k, Real &accumulator) {
+              const int &&ilev = k / VECTOR_SIZE;
```

Comment:
This is a case where casting the view to Real* from Scalar* would save ilev, vlev div-mod arithmetic and make the code straightforward.

Reply:
Fair. I'm just somewhat wary of casts like that; I really wish Kokkos had a reshape method with safety to ensure we don't make bad casts. The vector packs admittedly make this more difficult, but we could conceivably make the packs we have more general to avoid the `Pack -> Real` issue

Reply:
For safety, could make a cast wrapper that static_asserts on packs that don't support it. Tbc, we want to avoid runtime arithmetic that can be avoided.